src/quic/node_quic_buffer.cc

```diff
@@ -0,0 +1,169 @@
+#include "node_quic_buffer-inl.h"
+#include "util.h"
+#include "uv.h"
+
+#include <array>
+#include <algorithm>
+#include <functional>
+#include <vector>
+
+namespace node {
+namespace quic {
+
+namespace {
+inline bool IsEmptyBuffer(const uv_buf_t& buf) {
+  return buf.len == 0 || buf.base == nullptr;
+}
+}  // namespace
+
+QuicBuffer& QuicBuffer::operator+=(QuicBuffer&& src) noexcept {
+  if (tail_ == nullptr) {
+    // If this thing is empty, just do a move...
+    return *this = std::move(src);
+  }
+
+  tail_->next_ = std::move(src.root_);
+  // If head_ is null, then it had been read to the
+  // end, set the new head_ equal to the appended
+  // root.
+  if (head_ == nullptr)
+    head_ = tail_->next_.get();
+  tail_ = src.tail_;
+  length_ += src.length_;
+  rlength_ += src.length_;
+  size_ += src.size_;
+  count_ += src.size_;
+  Reset(&src);
+  return *this;
+}
+
+size_t QuicBuffer::Push(uv_buf_t* bufs, size_t nbufs, done_cb done) {
+  size_t len = 0;
+  if (nbufs == 0 || bufs == nullptr || IsEmptyBuffer(bufs[0])) {
+    done(0);
+    return 0;
+  }
+  size_t n = 0;
+  while (nbufs > 1) {
+    if (!IsEmptyBuffer(bufs[n])) {
+      Push(bufs[n]);
+      length_ += bufs[n].len;
+      rlength_ += bufs[n].len;
+      len += bufs[n].len;
+    }
+    n++;
+    nbufs--;
+  }
+  length_ += bufs[n].len;
+  rlength_ += bufs[n].len;
+  len += bufs[n].len;
+  Push(bufs[n], done);
+  return len;
+}
+
+void QuicBuffer::Push(std::unique_ptr<QuicBufferChunk> chunk) {
+  size_++;
+  count_++;
+  if (!tail_) {
+    root_ = std::move(chunk);
+    head_ = tail_ = root_.get();
+  } else {
+    tail_->next_ = std::move(chunk);
+    tail_ = tail_->next_.get();
+    if (!head_)
+      head_ = tail_;
+  }
+}
+
+size_t QuicBuffer::SeekHead(size_t amount) {
+  size_t n = 0;
+  size_t amt = amount;
+  while (head_ != nullptr && amt > 0) {
+    head_ = head_->next_.get();
+    n++;
+    amt--;
+    count_--;
+    rlength_ -= head_ == nullptr ? 0 : head_->buf_.len;
+  }
+  return n;
+}
+
+void QuicBuffer::SeekHeadOffset(ssize_t amount) {
+  // If amount is negative, then we use the full length
+  size_t amt = UNLIKELY(amount < 0) ?
+      length_ :
+      std::min(length_, static_cast<size_t>(amount));
+  while (head_ && amt > 0) {
+    size_t len = head_->buf_.len - head_->roffset_;
+    // If the remaining length in the head is greater than the
+    // amount we're seeking, just adjust the roffset
+    if (len > amt) {
+      head_->roffset_ += amt;
+      rlength_ -= amt;
+      break;
+    }
+    // Otherwise, decrement the amt and advance the read head
+    // one space and iterate from there.
+    amt -= len;
+    rlength_ -= len;
+    head_ = head_->next_.get();
+  }
+}
+
+size_t QuicBuffer::DrainInto(
+    add_fn add_to_list,
```

Comment:
Any reason to switch away from the template variant? The function isn’t stored anywhere, so using `std::function` will have unnecessary overhead, 1) in terms of an extra temporary heap allocation and, more importantly, 2) because the compiler can’t inline it anymore

Reply:
Compiler errors from the calls to it from the other inline DrainInto variants (unresolved symbol). I'll tweak it tho.